{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ac6f3135",
   "metadata": {},
   "source": [
    "<h1>Generate table from collection of phenopackets</h1>\n",
    "<p>A common task for the analysis of a cohort of individuals with pathogenic variants in a given gene is to generate a table with a summary of the findings. The pyphetool package has functionality to ingest a collection of phenopakcets and to generate several different kinds of tables that may be useful for publications of supplementary material sections.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "081a7501",
   "metadata": {},
   "outputs": [],
   "source": [
    "import phenopackets as php\n",
    "from google.protobuf.json_format import MessageToDict, MessageToJson\n",
    "from google.protobuf.json_format import Parse, ParseDict\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_colwidth', None) # show entire column contents, important!\n",
    "from collections import defaultdict\n",
    "import os\n",
    "import sys\n",
    "\n",
    "from phenopackets import Phenopacket\n",
    "from google.protobuf.json_format import Parse\n",
    "import json\n",
    "\n",
    "sys.path.insert(0, os.path.abspath('../../pyphetools'))\n",
    "from pyphetools.output import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e4e2f138",
   "metadata": {},
   "outputs": [],
   "source": [
    "phenopacket_dir = \"/Users/robinp/GIT/hrmd/notebooks/TRPM3/phenopackets\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1dfca97d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We extracted 17 GA4GH phenopackets\n"
     ]
    }
   ],
   "source": [
    "## NOT NEEDED ANYMORE\n",
    "phenopacket_paths = []\n",
    "if not os.path.isdir(phenopacket_dir):\n",
    "    raise ValueError(f\"{phenopacket_dir} was not a directory\")\n",
    "for root, dirs, files in os.walk(phenopacket_dir):\n",
    "    for file in files:\n",
    "        if file.endswith(\".json\"):\n",
    "            phenopacket_paths.append(os.path.join(root,file))\n",
    "print(f\"We extracted {len(phenopacket_paths)} GA4GH phenopackets\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e6842179",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "subject_or_biosample_id: \"Individual 5\"\n",
      "interpretation_status: CAUSATIVE\n",
      "variant_interpretation {\n",
      "  variation_descriptor {\n",
      "    gene_context {\n",
      "      value_id: \"HGNC:17992\"\n",
      "      symbol: \"TRPM3\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.c\"\n",
      "      value: \"NM_020952.6:c.2509G>A\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.g\"\n",
      "      value: \"NC_000009.12:g.70598463C>T\"\n",
      "    }\n",
      "    vcf_record {\n",
      "      genome_assembly: \"hg38\"\n",
      "      chrom: \"chr9\"\n",
      "      pos: 70598463\n",
      "      ref: \"C\"\n",
      "      alt: \"T\"\n",
      "    }\n",
      "    molecule_context: genomic\n",
      "  }\n",
      "}\n",
      "\n",
      "subject_or_biosample_id: \"1\"\n",
      "interpretation_status: CAUSATIVE\n",
      "variant_interpretation {\n",
      "  variation_descriptor {\n",
      "    gene_context {\n",
      "      value_id: \"HGNC:17992\"\n",
      "      symbol: \"TRPM3\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.c\"\n",
      "      value: \"NM_020952.6:c.2509G>A\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.g\"\n",
      "      value: \"NC_000009.12:g.70598463C>T\"\n",
      "    }\n",
      "    vcf_record {\n",
      "      genome_assembly: \"hg38\"\n",
      "      chrom: \"chr9\"\n",
      "      pos: 70598463\n",
      "      ref: \"C\"\n",
      "      alt: \"T\"\n",
      "    }\n",
      "    molecule_context: genomic\n",
      "    allelic_state {\n",
      "      id: \"GENO:0000135\"\n",
      "      label: \"heterozygous\"\n",
      "    }\n",
      "  }\n",
      "}\n",
      "\n",
      "subject_or_biosample_id: \"Individual 4\"\n",
      "interpretation_status: CAUSATIVE\n",
      "variant_interpretation {\n",
      "  variation_descriptor {\n",
      "    gene_context {\n",
      "      value_id: \"HGNC:17992\"\n",
      "      symbol: \"TRPM3\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.c\"\n",
      "      value: \"NM_020952.6:c.2509G>A\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.g\"\n",
      "      value: \"NC_000009.12:g.70598463C>T\"\n",
      "    }\n",
      "    vcf_record {\n",
      "      genome_assembly: \"hg38\"\n",
      "      chrom: \"chr9\"\n",
      "      pos: 70598463\n",
      "      ref: \"C\"\n",
      "      alt: \"T\"\n",
      "    }\n",
      "    molecule_context: genomic\n",
      "  }\n",
      "}\n",
      "\n",
      "subject_or_biosample_id: \"5-year-old girl\"\n",
      "interpretation_status: CAUSATIVE\n",
      "variant_interpretation {\n",
      "  variation_descriptor {\n",
      "    gene_context {\n",
      "      value_id: \"HGNC:17992\"\n",
      "      symbol: \"TRPM3\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.c\"\n",
      "      value: \"NM_020952.6:c.2509G>A\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.g\"\n",
      "      value: \"NC_000009.12:g.70598463C>T\"\n",
      "    }\n",
      "    vcf_record {\n",
      "      genome_assembly: \"hg38\"\n",
      "      chrom: \"chr9\"\n",
      "      pos: 70598463\n",
      "      ref: \"C\"\n",
      "      alt: \"T\"\n",
      "    }\n",
      "    molecule_context: genomic\n",
      "  }\n",
      "}\n",
      "\n",
      "subject_or_biosample_id: \"Individual 3\"\n",
      "interpretation_status: CAUSATIVE\n",
      "variant_interpretation {\n",
      "  variation_descriptor {\n",
      "    gene_context {\n",
      "      value_id: \"HGNC:17992\"\n",
      "      symbol: \"TRPM3\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.c\"\n",
      "      value: \"NM_020952.6:c.2509G>A\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.g\"\n",
      "      value: \"NC_000009.12:g.70598463C>T\"\n",
      "    }\n",
      "    vcf_record {\n",
      "      genome_assembly: \"hg38\"\n",
      "      chrom: \"chr9\"\n",
      "      pos: 70598463\n",
      "      ref: \"C\"\n",
      "      alt: \"T\"\n",
      "    }\n",
      "    molecule_context: genomic\n",
      "  }\n",
      "}\n",
      "\n",
      "subject_or_biosample_id: \"6\"\n",
      "interpretation_status: CAUSATIVE\n",
      "variant_interpretation {\n",
      "  variation_descriptor {\n",
      "    gene_context {\n",
      "      value_id: \"HGNC:17992\"\n",
      "      symbol: \"TRPM3\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.c\"\n",
      "      value: \"NM_020952.6:c.2509G>A\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.g\"\n",
      "      value: \"NC_000009.12:g.70598463C>T\"\n",
      "    }\n",
      "    vcf_record {\n",
      "      genome_assembly: \"hg38\"\n",
      "      chrom: \"chr9\"\n",
      "      pos: 70598463\n",
      "      ref: \"C\"\n",
      "      alt: \"T\"\n",
      "    }\n",
      "    molecule_context: genomic\n",
      "    allelic_state {\n",
      "      id: \"GENO:0000135\"\n",
      "      label: \"heterozygous\"\n",
      "    }\n",
      "  }\n",
      "}\n",
      "\n",
      "subject_or_biosample_id: \"7\"\n",
      "interpretation_status: CAUSATIVE\n",
      "variant_interpretation {\n",
      "  variation_descriptor {\n",
      "    gene_context {\n",
      "      value_id: \"HGNC:17992\"\n",
      "      symbol: \"TRPM3\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.c\"\n",
      "      value: \"NM_020952.6:c.2509G>A\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.g\"\n",
      "      value: \"NC_000009.12:g.70598463C>T\"\n",
      "    }\n",
      "    vcf_record {\n",
      "      genome_assembly: \"hg38\"\n",
      "      chrom: \"chr9\"\n",
      "      pos: 70598463\n",
      "      ref: \"C\"\n",
      "      alt: \"T\"\n",
      "    }\n",
      "    molecule_context: genomic\n",
      "    allelic_state {\n",
      "      id: \"GENO:0000135\"\n",
      "      label: \"heterozygous\"\n",
      "    }\n",
      "  }\n",
      "}\n",
      "\n",
      "subject_or_biosample_id: \"Individual 2\"\n",
      "interpretation_status: CAUSATIVE\n",
      "variant_interpretation {\n",
      "  variation_descriptor {\n",
      "    gene_context {\n",
      "      value_id: \"HGNC:17992\"\n",
      "      symbol: \"TRPM3\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.c\"\n",
      "      value: \"NM_020952.6:c.2509G>A\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.g\"\n",
      "      value: \"NC_000009.12:g.70598463C>T\"\n",
      "    }\n",
      "    vcf_record {\n",
      "      genome_assembly: \"hg38\"\n",
      "      chrom: \"chr9\"\n",
      "      pos: 70598463\n",
      "      ref: \"C\"\n",
      "      alt: \"T\"\n",
      "    }\n",
      "    molecule_context: genomic\n",
      "  }\n",
      "}\n",
      "\n",
      "subject_or_biosample_id: \"8\"\n",
      "interpretation_status: CAUSATIVE\n",
      "variant_interpretation {\n",
      "  variation_descriptor {\n",
      "    gene_context {\n",
      "      value_id: \"HGNC:17992\"\n",
      "      symbol: \"TRPM3\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.c\"\n",
      "      value: \"NM_020952.6:c.2810C>A\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.g\"\n",
      "      value: \"NC_000009.12:g.70553229G>T\"\n",
      "    }\n",
      "    vcf_record {\n",
      "      genome_assembly: \"hg38\"\n",
      "      chrom: \"chr9\"\n",
      "      pos: 70553229\n",
      "      ref: \"G\"\n",
      "      alt: \"T\"\n",
      "    }\n",
      "    molecule_context: genomic\n",
      "    allelic_state {\n",
      "      id: \"GENO:0000135\"\n",
      "      label: \"heterozygous\"\n",
      "    }\n",
      "  }\n",
      "}\n",
      "\n",
      "subject_or_biosample_id: \"4\"\n",
      "interpretation_status: CAUSATIVE\n",
      "variant_interpretation {\n",
      "  variation_descriptor {\n",
      "    gene_context {\n",
      "      value_id: \"HGNC:17992\"\n",
      "      symbol: \"TRPM3\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.c\"\n",
      "      value: \"NM_020952.6:c.2509G>A\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.g\"\n",
      "      value: \"NC_000009.12:g.70598463C>T\"\n",
      "    }\n",
      "    vcf_record {\n",
      "      genome_assembly: \"hg38\"\n",
      "      chrom: \"chr9\"\n",
      "      pos: 70598463\n",
      "      ref: \"C\"\n",
      "      alt: \"T\"\n",
      "    }\n",
      "    molecule_context: genomic\n",
      "    allelic_state {\n",
      "      id: \"GENO:0000135\"\n",
      "      label: \"heterozygous\"\n",
      "    }\n",
      "  }\n",
      "}\n",
      "\n",
      "subject_or_biosample_id: \"Individual 1\"\n",
      "interpretation_status: CAUSATIVE\n",
      "variant_interpretation {\n",
      "  variation_descriptor {\n",
      "    gene_context {\n",
      "      value_id: \"HGNC:17992\"\n",
      "      symbol: \"TRPM3\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.c\"\n",
      "      value: \"NM_020952.6:c.2509G>A\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.g\"\n",
      "      value: \"NC_000009.12:g.70598463C>T\"\n",
      "    }\n",
      "    vcf_record {\n",
      "      genome_assembly: \"hg38\"\n",
      "      chrom: \"chr9\"\n",
      "      pos: 70598463\n",
      "      ref: \"C\"\n",
      "      alt: \"T\"\n",
      "    }\n",
      "    molecule_context: genomic\n",
      "  }\n",
      "}\n",
      "\n",
      "subject_or_biosample_id: \"5\"\n",
      "interpretation_status: CAUSATIVE\n",
      "variant_interpretation {\n",
      "  variation_descriptor {\n",
      "    gene_context {\n",
      "      value_id: \"HGNC:17992\"\n",
      "      symbol: \"TRPM3\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.c\"\n",
      "      value: \"NM_020952.6:c.2509G>A\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.g\"\n",
      "      value: \"NC_000009.12:g.70598463C>T\"\n",
      "    }\n",
      "    vcf_record {\n",
      "      genome_assembly: \"hg38\"\n",
      "      chrom: \"chr9\"\n",
      "      pos: 70598463\n",
      "      ref: \"C\"\n",
      "      alt: \"T\"\n",
      "    }\n",
      "    molecule_context: genomic\n",
      "    allelic_state {\n",
      "      id: \"GENO:0000135\"\n",
      "      label: \"heterozygous\"\n",
      "    }\n",
      "  }\n",
      "}\n",
      "\n",
      "subject_or_biosample_id: \"2\"\n",
      "interpretation_status: CAUSATIVE\n",
      "variant_interpretation {\n",
      "  variation_descriptor {\n",
      "    gene_context {\n",
      "      value_id: \"HGNC:17992\"\n",
      "      symbol: \"TRPM3\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.c\"\n",
      "      value: \"NM_020952.6:c.2509G>A\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.g\"\n",
      "      value: \"NC_000009.12:g.70598463C>T\"\n",
      "    }\n",
      "    vcf_record {\n",
      "      genome_assembly: \"hg38\"\n",
      "      chrom: \"chr9\"\n",
      "      pos: 70598463\n",
      "      ref: \"C\"\n",
      "      alt: \"T\"\n",
      "    }\n",
      "    molecule_context: genomic\n",
      "    allelic_state {\n",
      "      id: \"GENO:0000135\"\n",
      "      label: \"heterozygous\"\n",
      "    }\n",
      "  }\n",
      "}\n",
      "\n",
      "subject_or_biosample_id: \"Individual 7\"\n",
      "interpretation_status: CAUSATIVE\n",
      "variant_interpretation {\n",
      "  variation_descriptor {\n",
      "    gene_context {\n",
      "      value_id: \"HGNC:17992\"\n",
      "      symbol: \"TRPM3\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.c\"\n",
      "      value: \"NM_020952.6:c.2509G>A\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.g\"\n",
      "      value: \"NC_000009.12:g.70598463C>T\"\n",
      "    }\n",
      "    vcf_record {\n",
      "      genome_assembly: \"hg38\"\n",
      "      chrom: \"chr9\"\n",
      "      pos: 70598463\n",
      "      ref: \"C\"\n",
      "      alt: \"T\"\n",
      "    }\n",
      "    molecule_context: genomic\n",
      "  }\n",
      "}\n",
      "\n",
      "subject_or_biosample_id: \"Individual 6\"\n",
      "interpretation_status: CAUSATIVE\n",
      "variant_interpretation {\n",
      "  variation_descriptor {\n",
      "    gene_context {\n",
      "      value_id: \"HGNC:17992\"\n",
      "      symbol: \"TRPM3\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.c\"\n",
      "      value: \"NM_020952.6:c.2509G>A\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.g\"\n",
      "      value: \"NC_000009.12:g.70598463C>T\"\n",
      "    }\n",
      "    vcf_record {\n",
      "      genome_assembly: \"hg38\"\n",
      "      chrom: \"chr9\"\n",
      "      pos: 70598463\n",
      "      ref: \"C\"\n",
      "      alt: \"T\"\n",
      "    }\n",
      "    molecule_context: genomic\n",
      "  }\n",
      "}\n",
      "\n",
      "subject_or_biosample_id: \"7-year-old patient\"\n",
      "interpretation_status: CAUSATIVE\n",
      "variant_interpretation {\n",
      "  variation_descriptor {\n",
      "    gene_context {\n",
      "      value_id: \"HGNC:17992\"\n",
      "      symbol: \"TRPM3\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.c\"\n",
      "      value: \"NM_020952.6:c.2509G>A\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.g\"\n",
      "      value: \"NC_000009.12:g.70598463C>T\"\n",
      "    }\n",
      "    vcf_record {\n",
      "      genome_assembly: \"hg38\"\n",
      "      chrom: \"chr9\"\n",
      "      pos: 70598463\n",
      "      ref: \"C\"\n",
      "      alt: \"T\"\n",
      "    }\n",
      "    molecule_context: genomic\n",
      "  }\n",
      "}\n",
      "\n",
      "subject_or_biosample_id: \"3\"\n",
      "interpretation_status: CAUSATIVE\n",
      "variant_interpretation {\n",
      "  variation_descriptor {\n",
      "    gene_context {\n",
      "      value_id: \"HGNC:17992\"\n",
      "      symbol: \"TRPM3\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.c\"\n",
      "      value: \"NM_020952.6:c.2509G>A\"\n",
      "    }\n",
      "    expressions {\n",
      "      syntax: \"hgvs.g\"\n",
      "      value: \"NC_000009.12:g.70598463C>T\"\n",
      "    }\n",
      "    vcf_record {\n",
      "      genome_assembly: \"hg38\"\n",
      "      chrom: \"chr9\"\n",
      "      pos: 70598463\n",
      "      ref: \"C\"\n",
      "      alt: \"T\"\n",
      "    }\n",
      "    molecule_context: genomic\n",
      "    allelic_state {\n",
      "      id: \"GENO:0000135\"\n",
      "      label: \"heterozygous\"\n",
      "    }\n",
      "  }\n",
      "}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ingestor = PhenopacketIngestor(indir=phenopacket_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "dfa9ae2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We got 17 phenopackets\n"
     ]
    }
   ],
   "source": [
    "patient_d = ingestor.get_patient_dictionary()\n",
    "print(f\"We got {len(patient_d)} phenopackets\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "689a265a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "782406e6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "venv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
